### ISSUE TITLE
Suggestion: Multiple devices

### ISSUE THREAD

**Issue by SafwatHalaby** [irungentoo/toxcore#843]
The optimal location for this suggestion is the Wiki.
But I do not have access to that, so I decided to post this here.
Feel free to move it and close this issue.

This is an alternative proposal for the one at http://wiki.tox.im/Multiple_Devices.
A key feature is the fact that it does not require a change to the Tox network protocol and a client (such as Venom) can implement this.

**_Summary:**_
Instead of creating 2 identities and linking them, the same identity is shared over the 2 devices transparently. This reduces the technical complexity a lot, and reduces user confusion: Each user will have only one public key to share with his friends.

**_How it works**_:

**Linking the devices (Sharing the identity):**
Alice already has an identity - which is a public/private key pair - on her PC, Alice decides to use her smartphone with Tox. Alice clicks a "Add an existing identity" button on the smartphone. Alice then inputs her public key. In the background, Alice's smartphone creates a temporarily identity and contacts Alice's PC via DHT. It then sends a "linking request", Alice accepts that request on her PC (We could make her type a secret phrase on both ends for security). Once the request is accepted, the Private/Public key pair is transferred - securely, either via DHT or p2p - from Alice's PC to Alice's smartphone. Alice's smartphone discards the temporarily identity and adopts the newly received one.
Alice's smartphone then "goes online", see next section.

**When a device goes online - Cooperation between devices sharing the same key:**
When a devices starts up / finishes the step above, it checks if another device is already online with the same public key (Scanning the DHT possibly requires a temporarily identity, correct me here). 2 Things can occur:

_A: There is a device already online with that same public key:_
We'll call the device which is already online the "master" (the PC in Alice's case).
We'll call the newly connecting device a "slave" (the smartphone in Alice's case).

The slave contacts the master via DHT (via a temporarily identity), and asks to be synced. The slave can easily prove it has possession of the same private key as the master by completing a challenge. Once this is proven, the master and the slave sync up: The slave(Smartphone) initiates a peer-to-peer connection to the master(PC), it then **opts out of the DHT**. Now, the phone is merely a "remote interface" to the PC, which is the actual Tox peer: Any action Alice performs on the slave(smartphone) is sent to the master (PC), only the master contacts the DHT and/or Alice's friends. Any action the PC receives from the DHT/from alice's friends is sent to the smartphone.

To sum it up, the slave is a "remote session" of the master.

_B: There is no device already online with that same public key:_
In that case, the newly connecting devices connects normally using the normal Tox procedure, no science here. The devices then becomes the "master".

**Some remarks:**
1. This also works with multiple slaves. The master should randomly choose a "top slave" among all the connected slaves, in case the master disconnects, the "top slave" becomes the master and the remaining slaves reconnect to their new master. 
2. The linking phase is really easy and already partially implemented: It is very similar to accepting a friend request.
3. The entire procedure can be implemented via a client and does not require changes to the Tox network protocol. (But it's best to implement in Tox Core)

**Comment by Oddwarg** [comment]
Tox is the most promising Skype alternative I have found to date, but this issue is the deal breaker. Casual users will continue to refuse to switch to Tox until it becomes easy to share some kind of identity across multiple devices. Why is there still no development on this front?

**Comment by GrayHatter** [comment]
@Oddwarg because no one has a solution that doesn't pose a security risk. Every workable solution I've seen either requires you trust some central authority, or allows for impersonation. Both of which negates the point of Tox existing.

**Comment by ProMcTagonist** [comment]
What? What's the problem with adding a new device as a friend, specifying that you want to connect the profile, putting in some PIN from one device on the other, and then updating your friends to send messages to all your devices?

**Comment by GrayHatter** [comment]
I don't understand how that would even work.

**Comment by ProMcTagonist** [comment]
Alice uses Tox Profile A to talk to Bob. 

Alice gets a phone and installs Antox on it, generating Tox Profile B.

Alice goes into her settings on Tox Profile A and puts B's ID into the Add Device field. 

A sends a special friend request to B, and A displays a 4-digit PIN.

B receives the request and pops up a dialog asking for the PIN.

When the PIN is correctly entered, A and B are connected and begin syncing message history, friends, etc using file transfers. A broadcasts to all contacts that it is also reachable at B now. Contacts' clients store this information however the multidevice spec says to.

Now when Bob sends a message to Alice it goes to both of her devices, which may themselves negotiate when each should alert or be silent (the way iMessage is supposed to be silenced on a phone when you're near your laptop).

Now tell me all the problems.

**Comment by Jeeppler** [comment]
@GrayHatter @ProMcTagonist  idea is very similar to what firefox sync uses or bluetooth for device paring.

**Comment by GrayHatter** [comment]
> Now tell me all the problems.

4 digit pin is insecure.

How do you deauth a device you lose?

What's to stop a user from hijacking your phone, adding a new device,
returning your phone?

How do clients know what to send to who where and when? (Like if a device
is offline, then comes back online)

How do you handle message receipts?

**Comment by ProMcTagonist** [comment]
>4 digit pin is insecure.

Maybe a few dictionary words in the user's language, then? Or just a longer PIN formatted like a phone number, because we're good at keeping those in short term memory for the few seconds we need. 

>How do you deauth a device you lose?

You would go into the settings from any authed device and deauth the chosen one. Then the remaining authed devices would update and tell contacts to stop talking to the deauthed one.

This is an area that needs more thought because an attacker could get ahold of one of your devices and deauth the rest. But maybe the devices could detect a one-to-many deauth and decide to require more PINs, or contacts could be involved, or something. It needs thought, but I'm pretty sure it can be solved.

>What's to stop a user from hijacking your phone, adding a new device, returning your phone?

You would be able to see all devices you have added, and notice if someone's added a sketchy one. Notifications should definitely appear on all devices when a new one is added. 

Also, physical access is physical access.

>How do clients know what to send to who where and when? (Like if a device is offline, then comes back online)

I'm not sure I understand properly, apologies. But if you mean what if one of Alice's devices goes offline mid-conversation then comes back, then it would receive new messages and sync the missed ones from another one of Alice's devices that was online the whole time. If no devices were on then it's the same as it is now with faux or real offline messaging.

>How do you handle message receipts?

I'm not sure I understand what you mean here either. Confusion about what constitutes as delivered or read when there are multiple devices involved? In that case we should definitely look to programs already handling this for inspiration and lessons, like Skype, iMessage, etc.

**Comment by ameenross** [comment]
I have a couple of ideas I'd like to throw out there.

It doesn't sound like a good idea to have the client talk to each of the devices of a friend. It should only have to communicate with a single device (after which the devices sync messages on demand). The selection of which device could be automatic, something arbitrary like type of device (desktop > laptop > tablet > phone > wearable) and weighted by relation age (maliciously added device then wouldn't be a MITM risk, unless not detected for a looooong time). Also, clients should visibly show the various devices of both the user *and* contacts, prompt the user when a device is added (desktop client should show that a phone was added when it was done via a laptop) or removed and probably have the option to prompt when a friend adds a device.

We're probably not there yet by a long shot, just trying to work in that direction.

**Comment by ProMcTagonist** [comment]
>It should only have to communicate with a single device (after which the devices sync messages on demand).

Then if that one goes offline friends won't know about other devices. Not acceptable user experience.

>Also, clients should visibly show the various devices of both the user and contacts, prompt the user when a device is added (desktop client should show that a phone was added when it was done via a laptop) or removed and probably have the option to prompt when a friend adds a device.

+1

**Comment by ameenross** [comment]
> Then if that one goes offline friends won't know about other devices. Not acceptable user experience.

I didn't mean that clients shouldn't know about the various devices of friends, rather the contrary. But it should select one of the online devices to send messages to. The receiver has to do message synching anyway, it's inevitable. And I don't think people would be happy if they have to send that PDF/jpg/whatever 5 times just because their contact is online on 5 devices.

**Comment by ProMcTagonist** [comment]
Ahh, that's a good idea :)

**Comment by weedy** [comment]
From a mobile standpoint it's obvious that I would want to nominate my desktop/laptop/NAS as the primary ID.
And then sync to that, over TCP.

**Comment by GrayHatter** [comment]
Client A installs antox on her phone, then uTox on her computer, and syncs to her account. Her friend Client B accepts Antox, and uTox as her devices. I steal her phone, DOS her uTox from the tox network, and add my malicious client to her account because I have her phone. Her friends accepts my malicious client as her friend, I keep randomly DOSing her so she never knows that the messages she thinks she's getting from her friends are actually from me.

**Comment by ameenross** [comment]
I'm not quite sure I understand that scenario. Could you stick to the classic Alice and Bob convention and use bullet points?

**Comment by JacobHenner** [comment]
I'm still learning more about the protocol, so I apologize in advance if any of this is naive:

Can we split this into two problems? The first being message synchronization between devices owned by a user, and the second being the process the proper direction of messages to a user who is running Tox on multiple devices (either concurrently, or at discrete intervals). 

For the first problem, perhaps the user could (in each client), specify the identities of the other clients he/she uses. When one identity receives a message, it attempts to communicate with the other identities. If they are online, the clients send a confirmation that the message was globally received. If it was not (in some reasonable period of time), the clients who did not receive the message request it from the clients who did, and it is synced. If the others (some or all) are offline, they request missed messages (perhaps using an identifier of the last received message) from online clients. If there are no online clients, the client waits until one becomes available. Obviously this has some drawbacks, including:
* Each device must have the addresses of the other devices in order to sync efficiently.
   * We also must avoid confirmation loops, where device A, B and C receive a message M, A asks B and C if it is received, B asks C if it was received, C asks B if it was received, C and B then ask A if it was received, etc.
* There will be additional traffic overhead.
* If a client is to sync messages in the future with a client which was offline/inaccessible at the time the message was received, it must store the message. This is not an issue if the device it is stored on is secure, but that's not a realistic assumption, and although the user might enable logging through the UI he/she uses, this would require logging at the protocol level. 
    * If this protocol-level storage is enabled, how long are messages stored before they are deleted? 
        * The messages could be deleted from this log as soon as all of the specified clients have acknowledged receipt of it, still non-optimal

For the second, could some sort of revocable signature be stored in the DHT? A client can sign other identities as theirs (just as they specify these identities, see above), and when an identity is requested, the DHT returns all associated? This would require all linked identities to have signed each other. A revocation from any associated client would void the connection, and messages would only be sent the the identity of the original friend request.
Alternatively, the signing can occur, and then the linked clients can send the signatures to all contacts on the friends list. Revocation would work similarly. Of course, there would be a bit more lag here than if the info was simply maintained in the DHT somehow.

IThiis is an important hurdle for adoption. Without multi-device support and synchronization, Tox is non-competitive for the average (read: slightly privacy conscious) user, imho.

**Comment by mightyCelu** [comment]
Why is it necessary to create new IDs for every device? Wouldn't the easiest way be to support a identity being used on multiple devices and having multiple addresses stored in the DHT? This would allow for a better usability because every user will be known by exactly one ID.

Of course, compromising a device would result in the identity being lost but if there is complete access to a device there is no definite way of distinguishing a attacker from the owner. Thus I think all methods to try and recover from a situation like this will be very complex without being able to solve the problem in the end. It is basicly the same problem as with PGP where the solution is to keep the key encrypted on every device.

As for which device among multiple to chose for sending to one could have a flag in the DHT and a special message for recent communication partners indicating the "primary device" that is sent every time the user interacts from a different device than the last one.

For video calls it also might be possible to request a handoff between two devices but this is a less important feature.

**Comment by JacobHenner** [comment]
> Why is it necessary to create new IDs for every device? Wouldn't the easiest way be to support a identity being used on multiple devices and having multiple addresses stored in the DHT? This would allow for a better usability because every user will be known by exactly one ID.

* In my proposal, they could still be "known" by one ID. The others would be transparently linked. 

> As for which device among multiple to chose for sending to one could have a flag in the DHT and a special message for recent communication partners indicating the "primary device" that is sent every time the user interacts from a different device than the last one.

* If an identity was connected to multiple addresses, and a device designated as the "primary device" disconnected (let's say there was a connectivity issue), how would the others then negotiate which is the primary, if neither are actively being used? What would happen to messages sent during this time? Would the other individual sending the messages to this user need to check the DHT each time he/she sends a message, to see which device is "primary"? How would they sync?

**Comment by mightyCelu** [comment]
>If an identity was connected to multiple addresses, and a device designated as the "primary device" disconnected (let's say there was a connectivity issue), how would the others then negotiate which is the primary, if neither are actively being used? What would happen to messages sent during this time? 

I think the issue with the primary device having connectivity issues is not that different from the situation that is present at the moment, the sender will detect that the device does not reply. It then can then send the message to one of the remaining devices (e.g. the last primary or a random one). The receiving device could then just declare itself as the primary node until the user uses on if his devices.
>Would the other individual sending the messages to this user need to check the DHT each time he/she sends a message, to see which device is "primary"? How would they sync?

For the first message the sending device has to look the address up in the DHT anyway so it will also retrieve the flag. When the primary device changes it can then contact all partners of ongoing conversations and inform about the change so they don't need to poll the DHT.

This dynamic behavior could also be configurable to prevent certain devices from being primary (mobilephones, maybe only if connected to wifi) or having a static primary node (server, desktop). The fallback to random or broadcast mode for sending devices would guarantee message delivery.

**Comment by JacobHenner** [comment]
> The receiving device could then just declare itself as the primary node until the user uses on if his devices.

So the new primary node declares itself primary because the other party randomly decided to send to it? Couldn't this be abused, such that an adversary always selects the non-active device, messages it, it claims it's the active, and now conflicts with the true active? How would the true active be able to refute this? How would the true active renegotiate its active flag with the other client claiming that it's active? 

This also still does not address message synchronization.



> For the first message the sending device has to look the address up in the DHT anyway so it will also retrieve the flag. When the primary device changes it can then contact all partners of ongoing conversations and inform about the change so they don't need to poll the DHT.

What if the primary device doesn't know it's about to change (i.e. user picks up smartphone as he walks out of his room).

**Comment by davidbenoit** [comment]
In a world of multiple devices per user, the concept of a "primary" device can't exist.  They need to be seen as equals.  Otherwise, you will constantly have to promote and demote nodes based on a user preference, deal with lost/damaged/compromised primary nodes, etc.  If each is a peer, thy all have the same rights.  Notify each peer whenever there is a new device, an exit device/peer needs to authorize it, can reject it, and can revoke anything from the list.  To get on the list, use a hash QR code, so that device need to be "close" (or some other explicitly-controlled-by-the-user mechanism to send it) to add a device.  That's what bleep does.

**Comment by ProMcTagonist** [comment]
"Primary" devices should be used solely to keep friends from being forced to send the same thing five times as was brought up earlier

**Comment by hypereddie** [comment]
> "Primary" devices should be used solely to keep friends from being forced to send the same thing five times as was brought up earlier
What if the GUI/client knew that an contact had multiple devices and when sending a message the GUI/client just sent a copy to each device, rather than the user having to do it manually?

**Comment by ProMcTagonist** [comment]
@hypereddie Yes, of course multiple device IDs would be transparent to end users. It's a bandwidth/cycles problem.

**Comment by hypereddie** [comment]
> It's a bandwidth/cycles problem.

Maybe I'm missing the concern..but you're gonna have to spend bandwidth somewhere to get the messages to each device ¯\_(ツ)_/¯

**Comment by ProMcTagonist** [comment]
@hypereddie 
>ameenross commented a day ago
>
 >>  Then if that one goes offline friends won't know about other devices. Not acceptable user experience.
>
>I didn't mean that clients shouldn't know about the various devices of friends, rather the contrary. But it should select one of the online devices to send messages to. The receiver has to do message synching anyway, it's inevitable. And I don't think people would be happy if they have to send that PDF/jpg/whatever 5 times just because their contact is online on 5 devices.

**Comment by JacobHenner** [comment]
> For the second, could some sort of revocable signature be stored in the DHT? A client can sign other identities as theirs (just as they specify these identities, see above), and when an identity is requested, the DHT returns all associated? This would require all linked identities to have signed each other. A revocation from any associated client would void the connection, and messages would only be sent the the identity of the original friend request.
> Alternatively, the signing can occur, and then the linked clients can send the signatures to all contacts on the friends list. Revocation would work similarly. Of course, there would be a bit more lag here than if the info was simply maintained in the DHT somehow.

Wouldn't this solve the issue of getting messages to the user? And since the rest of the burden would be on the synchronization side (which has some reasonable solutions), this should work...

**Comment by mightyCelu** [comment]
Okay, I think the idea of a primary device is not the best.
Regarding the signed multiple device IDs: How exactly would the revocation work? If every device have to sign each other's ID there are lots of opportunity to have a corrupted, maybe even fragmenting fragmenting the devices so that the friends are connected to different fragments.

**Comment by tge12345** [comment]
Another newbie contribution (didn't follow the discussion from the beginning, limited technical knowledge about tox):
What about having TWO tox IDs - one per user and one per device, so I personally am 12345 and my two devices are 23456 and 34567. When I log on, I register the current user/device ID pair(s). My friends only need to know 12345, their clients have to show only my user ID if there is at least one of my devices.
Then, afterwards the whole thing is just another group chat ... of course meaning, that contents has to be transmitted over the network <n> times, but with any kind of sync model this is necessary too (actually I think the whole discussion about saving network traffic is pointless - either my friend's client sends data <n> times, or my devices sync among them - in the end about the same, I'd guess).
No idea yet about key exchange etc, but maybe this yields some inspiration ...

**Comment by GrayHatter** [comment]
@tge12345 it's not a bad idea, we've talked about that style already. The problem with this one that no one's come up with a good solution to, is how to transfer around the private keys needed to decrypt all the messages.

**Comment by tge12345** [comment]
After reading #1100 and brain being close to overflow ...
@GrayHatter Somewhere there was a proposal that sounded quite logical to me, something like this (if understood correctly):
The only key to exchange would be the one of my user ID, and between my devices, and only once. And, the devices sit right in front of me while doing that. No such thing as master/slave or primary/secondary.
Device 23456 is already in use, let's call it initial, same as single-device. Now I want to add device 34567, perform a regular set-up on 34567, then I have two entirely independent devices in front of me (the new one has a new user ID too, 11111). Now I can securely communicate between both, because I'm both users and have both devices. From the initial device I send the profile data of 12345, on the new device I "accept and replace". Done.
Again, humble disclaimer: Of course it cannot be that simple, or somebody else would have come up with it already ...

**Comment by GrayHatter** [comment]
@tge12345 no, that wouldn't work, to decrypt packets you'd need the private key on every device you want to use. Forget anything about user_id. Think only of private and public keys. To read data you need the private key.

**Comment by tge12345** [comment]
@GrayHatter In general I believe I understand public key encryption, however I only have some rough idea how it is used in tox. Is there some doc (besides looking into the sources) anywhere explaining which keys are used where and how?

**Comment by tge12345** [comment]
Me again ... I looked through the docs that appeared relevant to me and into the sources. I'm far from understanding everything, but feel better now.
Now I have two aspects:
1) is there something like a structured requirement spec, something like:
  - what do we (and supposedly users) actually want when talking about 
    multi-device-support? Probably broken down into sub-requirements, 
    considering all perspectives, like end-user experience perspective 
    (usability), security, ...
  - for each such requirement, what are the solution options (pros, 
    cons, effort, ...)
  - all of that in a textual form and some summary table rating all 
    options, ultimately yielding the best solution

I believe that this reduces misunderstandings what we are actually talking about and hopefully provides some consensus (right now there is a set of lengthy discussion threads - hard to comprehend to newcomers like me :-)
I think that the wiki would be a good place to do this and everybody can nicely contribute.

2) Technically, I'm still pondering about that idea of separate user and device IDs, particularly (again, disclaimer: still limited knowledge):
- today there is a group chat function which allows to talk to set of 
  remote tox IDs
- keeping this very concept but replace "tox ID" by "device ID" would 
  not change anything so far - applicable to all communication and 
  encryption aspects
- Now linking the device IDs of a user together could be done on client 
  level (as described in #97). I never tried it but imagine 
  it fairly close to what I have in mind - the only disadvantage I see 
  is that all my peer friends have to group my devices individually 
  (assuming, I have multiple devices).
  If there were something like a "directory level" on top of the 
  device IDs in the DHT, e.g. "userID -> deviceID" (created whenever a 
  device starts up, the device must know the user ID) used purely as a 
  "name service" -> no need to transfer keys around (a user is just an 
  ID or even a textual name like "tge12345", no need to 
  worry about collisions as this is in the scope of my devices only).
  The userID could be part of the encrypted DHT entry of the device,
  so it is not visible anywhere -> nobody else would be able to 
  claim to be me (and even if - see below).
  The device-DHT-entry could even contain the list of 
  other devices belonging to a user, just for detecting imposters 
  (however that might be more complicated and not necessary)
- As a user I continue to give my friends all my device IDs, their 
  clients starts up, looks through their friends (now devices) to see 
  who is online and collect the userIDs
- The whole procedure would be completely optional - if I have only one 
  device, I register no userID and everything behaves as-is
- If I add a second device, I imagine something as posted earlier: 
  regular install new device (generates new keys and device ID), I 
  manually "send user data" from initial to new device and "accept" it 
  there (or just manually configure the "user ID" in the new client ...). The new device updates it's DHT entry with the new user ID.
- How to prevent that somebody else claims to be user "tge12345"? Not 
  needed at all, as long as my friends don't add any unknown device IDs 
  (as they are not supposed to today either), they will never see any 
  false tge12345 devices. As stated above, there is even nothing wrong 
  with multiple users with the same name/ID (only the clients have to 
  deal with that when displaying/grouping)

From the usability perspective this is slightly more cumbersome than today but appears to me acceptable - think of "adding a trusted device belonging to my friend" instead of just "adding a friend" (and once I have done that, I don't need to worry about multiple devices anymore).

Disadvantage: Clients must be able to handle that user/device grouping, but does not sound too complicated to me (and if they don't, they would see multiple devices not linked together -> no major harm).
In theory clients would/should even be able to expose the user/device relationship and for example, allow sending data explicitly to specific devices (I tell my friend before: "please send that large image to my laptop only")

**Comment by drequivalent** [comment]
It's been a year already. Any progress on that?

**Comment by GrayHatter** [comment]
https://docs.google.com/document/d/1op6zGR0KYdF7tTWSSX79KQieJu30vLZ6XG327kIBhxQ/edit

**Comment by SkyzohKey** [comment]
I came to a (maybe stupid) idea for multi devices.
Why not having 2 keypair. I mean, when the user create it's profile toxcore would generate an `account_keypair` and a `profile_keypair`. Then the profile would add the account via it's public_key. Another client could link to the "account" by adding it via the `account_pubkey`. When linked, the account would send it's private_key to the other device in order for it to also have a copy of the account (friends, etc). Of course, this would need to be secure.

ie.  `profiles[pn]` ↔ `account` ↔ `friends[fn]`
Where `fn` is the friends_number and `pn` the profiles_number.
Basically friends would add you using the `account_pubkey`. The the account part would send the request to all the connected and linked profiles. The same would apply for other actions (messages, audio, etc).

Well, this is just an idea i got and i think that it could works if well implemented. :)

**Comment by ProMcTagonist** [comment]
>the account would send it's private_key to the other device

pls no

**Comment by GrayHatter** [comment]
@SkyzohKey come talk with is in #tox-dev on freenode, that's a much better place for ideas, and comments

**Comment by William-Shadowruby** [comment]
How about a WOT model

First  profile-a would mark profile-b as being part of the same meta-profile

Profile-b then must do the same to profile-a.

If we want to add profile-c to this meta profile and vise versa.

profile-a and profile-b would both need to mark profile-c as being part of the meta profile

So on and so forth for the trivial implementation.

We could also make it so it only requires a majority of profiles to have marked another profile as part of the meta profile for it to be part of the meta profile.

Additionally profiles could have weights that influence how much their mark of a profile influences their acceptance into the meta-profile.

Finally if a profile is no longer marked by enough other profiles as being part of the meta-profile it should be removed from the profile.

I feel like this is a good structure because it seems theoretically more secure the more profiles you add because as you add devices it requires a greater amount of other profile to be authorized as part of the meta-profile.

**Comment by naught101** [comment]
Definitely do not use majority based profile merging.

For example:

- my grandmother creates a tox, so she can communicate securely with my family.
- evil cracker creates 3 tox accounts, merges them, and then claims that my grandmother's account is part of the same profile.
- my dad sends my grandmother his credit card details, thinking the line is secure, but the message actually goes to the evil cracker as well.

Merging should be explicitly requested from all parties. I think it would be good to have an "account X has claimed identity with your account." notice, and the ability to deny that claim, or to manually reciprocate. But I don't think it'd even be a good idea to have an "ok, reciprocate" button on the notice, because users often just press the yes/ok button without actually reading the message...

On 26 March 2016 6:49:34 am AEDT, William-Shadowruby <notifications@github.com> wrote:
>How about a WOT model
>
>First  profile-a would mark profile-b as being part of the same
>meta-profile
>
>Profile-b then must do the same to profile-a.
>
>If we want to add profile-c to this meta profile and vise versa.
>
>profile-a and profile-b would both need to mark profile-c as being part
>of the meta profile
>
>So on and so forth for the trivial implementation.
>
>We could also make it so it only requires a majority of profiles to
>have marked another profile as part of the meta profile for it to be
>part of the meta profile.
>
>Additionally profiles could have weights that influence how much their
>mark of a profile influences their acceptance into the meta-profile.
>
>Finally if a profile is no longer marked by enough other profiles as
>being part of the meta-profile it should be removed from the profile.
>
>I feel like this is a good structure because it seems theoretically
>more secure the more profiles you add because as you add devices it
>requires a greater amount of other profile to be authorized as part of
>the meta-profile.
>
>---
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/irungentoo/toxcore/issues/843#issuecomment-201456228

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by SkyzohKey** [comment]
The multidevice pairing should be done with a QR Code or a temporary generated key from the "main" profile, this would avoid any possibility an evil guy could know the key and avoid the user having to click yes/no ;)

**Comment by William-Shadowruby** [comment]
I've though about it a bit more and wrote a bit up

Each profile shall contain a meta-profile database.

This meta-profile database shall contain:
	* A List of other profiles in the meta profile
	* Each profile in this list should contain a list of the profiles that trust this profile

For a profile (hereby refered to as profile-b) to be added to the meta-profile database the following shall occur

First it shall send a request a preliminary accptance to the meta-profile to a profile (hereby refered to as profile-a) already within the meta-profile.

Then profile-a shall respond with a preliminary accept or preliminary deny.

Profile-b then will generate a n-length random non-unicode character string to be displayed to the user and send an acceptance confirmation to profile-a.

Then the user shall enter the same string into the device with profile-a whitch shall send it to profile-b.

Profile-b then compares the recived string from profile-a to the generated string. 

If the strings are identical profile-b sends a confirm responce to proflile-a.

Then profile-a shall also preform this process.

Then profile-a shall update its local meta-profile database with profile-b and broadcast that it now preliminalrily-trusts profile-b.

This process is repeated for each other profile that is part of the meta-profile.

Once each profile that is part of the meta profile preliminarly trust profile-b, profile-b becomes part of the meta-profile.


To remove a profile from a meta-profile each other profile shall first remove that profile from its list of trusted profiles then broadcast that it no longer trust that profile.

**Comment by ewtoombs** [comment]
Has any progress been made here? This is a huge gaping hole in tox's feature set.

**Comment by GrayHatter** [comment]
yes, I've merged v0.1.4 into my multidevice branch. And I'm working on hammering out all the remaining bugs.

I'll release something once it can create working save files again; right now, that's the only reason uTox doesn't support mdevice.

**Comment by thekyriarchy** [comment]
Can this report be officially closed then?

**Comment by akamaus** [comment]
@GrayHatter  What's the current status? Do you have plans to merging it into upstream?
